Journeyman/learned concept

journeyman-msft opened the pull request:
Using a learned selector for quanser qube.
![image](https://user-images.githubusercontent.com/52263960/124202506-566b8e80-da8f-11eb-8589-a9e8b6f30592.png)

Please confirm:
1. Training for the selector is similar (NPIL of 1mil was found to be needed) 
![image](https://user-images.githubusercontent.com/52263960/124202843-2cff3280-da90-11eb-9c3c-f14de8a8c23a.png)

2. Export brain work with visualizer using `python main.py --render --test-exported <PORT>`

3. automated test
```
pytest -s \
    --acr_name <ACR_NAME> \
    --brain_name <BRAIN_NAME> \
    --port <PORT> \
    --render False \
```

shnayder commented on Python/samples/quanser-qube/tests/test_multi_concept.py:
```diff
@@ -0,0 +1,72 @@
+"""
+Pytest for confirming multi concept brain functions with learned selector,
+one learned concept, and one programmed concept. Assessment with multi-concept 
```
what do you mean by "assessment with multi-concept is not available"? It should work fine to assess the selector concept, though testing with the exported brain seems perfectly fine too. Suggest just removing this sentence.

journeyman-msft replied:
I learned this on Friday. I misunderstood M5 was assessment for programmed concepts (not selectors). I can fix

shnayder commented:
I'm just about to try training this myself, but based on the code this all looks good.

akzaidi commented:
![image](https://user-images.githubusercontent.com/1307603/124840734-7f34cd80-df40-11eb-95c0-812d5892da57.png)

tried this yesterday and worked great. exported worked like a charm too. might not need as long as an NPIL, selector finished training after 1.2M and swingup after 500k.

exported results:

![output](https://user-images.githubusercontent.com/1307603/124841652-92489d00-df42-11eb-9d07-dbd338d32026.gif)

journeyman-msft commented:
> ![image](https://user-images.githubusercontent.com/1307603/124840734-7f34cd80-df40-11eb-95c0-812d5892da57.png)
> 
> tried this yesterday and worked great. exported worked like a charm too. might not need as long as an NPIL, selector finished training after 1.2M and swingup after 500k.
> 
> exported results:
> 
> ![output](https://user-images.githubusercontent.com/1307603/124841652-92489d00-df42-11eb-9d07-dbd338d32026.gif)

That's great to hear! Perhaps i'll lower the NPIL to 500k and see if it still works. Thanks for the gif too, i'll absorb that into the README.

journeyman-msft commented:
Interestingly, 500k was not enough
![image](https://user-images.githubusercontent.com/52263960/124845825-616d6580-df4c-11eb-8734-1da3b83b44cb.png)

shnayder commented:
My test also worked! Training the selector only took ~800K iterations this time:
![image](https://user-images.githubusercontent.com/530805/124826447-b3e95a80-df29-11eb-9cea-5d4ccff62bba.png)

@journeyman-msft  I ran into a bunch of python issues trying to get the sim to run locally. Traced it to pinned versions in requirements.txt — do you really need those exact versions? In particular scipy wouldn't build on my version of OS X until I removed the pin. And wouldn't you want latest bonsai-cli too?

shnayder commented:
Now I want to know what the decision boundary the selector learned looks like. Once the new exported brain APIs report which concept was chosen, we can make some plots  :)

shnayder commented:
@journeyman-msft two small notes: 
1. can you add the reference frame for the two state angles to the comments? (i.e. is 0 radians straight up or down?). 
2. I just saw Mike trying to create a brain from sim for qube and then saw errors about missing state vars—does the interface file promise some target values that aren't actually sent by the sim? (I haven't looked at the code to confirm—might have been some other issue)

journeyman-msft commented:
- removed target_ items in interface.json
- increased range of alpha for in interface.json
- added img for reference frame

![image](https://user-images.githubusercontent.com/52263960/126235329-3b2529b7-ed38-415e-a32e-f1fef4cddabd.png)
